### PR TITLE
Basic 1344

### PR DESCRIPTION
**Pull request by sherakama**
- One new step definition to allow for pystrings (multiple line input)
- Tests the stanford_tag_translate feature
- Updates the core course feature

### CODE REVIEW

**Comment by jbickar** [comment]
I get an exception with `stanford_courses_tag_translate.feature`:

```
spr-jbickar-2: ~/Documents/working/linky_clicky/jsa$ behat -p local -s all features/stanford_courses_tag_translate.feature -v
Feature: Stanford Courses Tag Translate
  In order to ensure that stanford courses tag translation works.
  As an administrative user
  I want to ensure that the Stanford Courses Tag Translate module is working properly

  Background:                                                    # features/stanford_courses_tag_translate.feature:6
    Given the "stanford_courses_tag_translate" module is enabled # SWSFeatureContext::theModuleIsEnabled()
      exception 'Exception' with message '
      ' in /Users/jbickar/Documents/working/linky_clicky/includes/bootstrap/SWSFeatureContext.php:82
      Stack trace:
      #0 [internal function]: SWSFeatureContext->theModuleIsEnabled('stanford_course...')
      #1 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Testwork/Call/Handler/RuntimeCallHandler.php(98): call_user_func_array(Array, Array)
      #2 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Testwork/Call/Handler/RuntimeCallHandler.php(54): Behat\Testwork\Call\Handler\RuntimeCallHandler->executeCall(Object(Behat\Behat\Definition\Call\DefinitionCall))
      #3 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Testwork/Call/CallCenter.php(125): Behat\Testwork\Call\Handler\RuntimeCallHandler->handleCall(Object(Behat\Behat\Definition\Call\DefinitionCall))
      #4 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Testwork/Call/CallCenter.php(80): Behat\Testwork\Call\CallCenter->handleCall(Object(Behat\Behat\Definition\Call\DefinitionCall))
      #5 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Behat/Tester/Runtime/RuntimeStepTester.php(125): Behat\Testwork\Call\CallCenter->makeCall(Object(Behat\Behat\Definition\Call\DefinitionCall))
      #6 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Behat/Tester/Runtime/RuntimeStepTester.php(73): Behat\Behat\Tester\Runtime\RuntimeStepTester->testDefinition(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\StepNode), Object(Behat\Behat\Definition\SearchResult), false)
      #7 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Behat/Hook/Tester/HookableStepTester.php(74): Behat\Behat\Tester\Runtime\RuntimeStepTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\StepNode), false)
      #8 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Behat/EventDispatcher/Tester/EventDispatchingStepTester.php(73): Behat\Behat\Hook\Tester\HookableStepTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\StepNode), false)
      #9 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Behat/Tester/StepContainerTester.php(59): Behat\Behat\EventDispatcher\Tester\EventDispatchingStepTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\StepNode), false)
      #10 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Behat/Tester/Runtime/RuntimeBackgroundTester.php(67): Behat\Behat\Tester\StepContainerTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\BackgroundNode), false)
      #11 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Behat/EventDispatcher/Tester/EventDispatchingBackgroundTester.php(73): Behat\Behat\Tester\Runtime\RuntimeBackgroundTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), false)
      #12 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Behat/Tester/Runtime/RuntimeScenarioTester.php(102): Behat\Behat\EventDispatcher\Tester\EventDispatchingBackgroundTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), false)
      #13 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Behat/Tester/Runtime/RuntimeScenarioTester.php(70): Behat\Behat\Tester\Runtime\RuntimeScenarioTester->testBackground(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), false)
      #14 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Behat/Hook/Tester/HookableScenarioTester.php(74): Behat\Behat\Tester\Runtime\RuntimeScenarioTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\ScenarioNode), false)
      #15 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Behat/EventDispatcher/Tester/EventDispatchingScenarioTester.php(103): Behat\Behat\Hook\Tester\HookableScenarioTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\ScenarioNode), false)
      #16 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Behat/Tester/Runtime/RuntimeFeatureTester.php(83): Behat\Behat\EventDispatcher\Tester\EventDispatchingScenarioTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\ScenarioNode), false)
      #17 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Behat/Hook/Tester/HookableFeatureTester.php(72): Behat\Behat\Tester\Runtime\RuntimeFeatureTester->test(Object(Behat\Behat\Context\Environment\UninitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), false)
      #18 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Behat/EventDispatcher/Tester/EventDispatchingFeatureTester.php(71): Behat\Behat\Hook\Tester\HookableFeatureTester->test(Object(Behat\Behat\Context\Environment\UninitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), false)
      #19 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Testwork/Tester/Runtime/RuntimeSuiteTester.php(63): Behat\Behat\EventDispatcher\Tester\EventDispatchingFeatureTester->test(Object(Behat\Behat\Context\Environment\UninitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), false)
      #20 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Testwork/Hook/Tester/HookableSuiteTester.php(73): Behat\Testwork\Tester\Runtime\RuntimeSuiteTester->test(Object(Behat\Behat\Context\Environment\UninitializedContextEnvironment), Object(Behat\Testwork\Specification\GroupedSpecificationIterator), false)
      #21 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Testwork/EventDispatcher/Tester/EventDispatchingSuiteTester.php(72): Behat\Testwork\Hook\Tester\HookableSuiteTester->test(Object(Behat\Behat\Context\Environment\UninitializedContextEnvironment), Object(Behat\Testwork\Specification\GroupedSpecificationIterator), false)
      #22 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Testwork/Tester/Runtime/RuntimeExercise.php(71): Behat\Testwork\EventDispatcher\Tester\EventDispatchingSuiteTester->test(Object(Behat\Behat\Context\Environment\UninitializedContextEnvironment), Object(Behat\Testwork\Specification\GroupedSpecificationIterator), false)
      #23 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Testwork/EventDispatcher/Tester/EventDispatchingExercise.php(70): Behat\Testwork\Tester\Runtime\RuntimeExercise->test(Array, false)
      #24 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Testwork/Ordering/OrderedExercise.php(80): Behat\Testwork\EventDispatcher\Tester\EventDispatchingExercise->test(Array, false)
      #25 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Testwork/Tester/Cli/ExerciseController.php(146): Behat\Testwork\Ordering\OrderedExercise->test(Array, false)
      #26 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Testwork/Tester/Cli/ExerciseController.php(108): Behat\Testwork\Tester\Cli\ExerciseController->testSpecifications(Object(Symfony\Component\Console\Input\ArgvInput), Array)
      #27 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Testwork/Cli/Command.php(63): Behat\Testwork\Tester\Cli\ExerciseController->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
      #28 /Users/jbickar/Documents/working/linky_clicky/vendor/symfony/console/Command/Command.php(256): Behat\Testwork\Cli\Command->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
      #29 /Users/jbickar/Documents/working/linky_clicky/vendor/symfony/console/Application.php(787): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
      #30 /Users/jbickar/Documents/working/linky_clicky/vendor/symfony/console/Application.php(186): Symfony\Component\Console\Application->doRunCommand(Object(Behat\Testwork\Cli\Command), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
      #31 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/src/Behat/Testwork/Cli/Application.php(121): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
      #32 /Users/jbickar/Documents/working/linky_clicky/vendor/symfony/console/Application.php(117): Behat\Testwork\Cli\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
      #33 /Users/jbickar/Documents/working/linky_clicky/vendor/behat/behat/bin/behat(31): Symfony\Component\Console\Application->run()
      #34 {main}
```

**Comment by jbickar** [comment]
Where are `includes/features/stanford_courses.feature` and `includes/features/stanford_courses_tag_translate.feature` called from? They do not appear to be symlinked in `jsa`.

**Comment by sherakama** [comment]
@jbickar, ensure the site you are testing on has the `stanford_courses_tag_translate` module available and can be enabled. That Given statement is not a new function and should not cause you an exception error. I suspect it may be in the Drupal side of things.

I have not symlinked the new tests as not everyone on our team is using the patched version of Drush required for this.

**Comment by jbickar** [comment]
Doh.

File: includes/features/stanford_courses_tag_translate.feature [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    Then I should be on "admin/config/stanford/courses/tag-translate"
+    Then I should not see "zzzzzzzZZZZZzzzzzz"
+
+  # This scenario only works if the import scenario about has been run
+  @api @content @dev @deploy
+  Scenario: Edit tag translations
+    Given I am logged in as a user with the "administrator" role
+    And I am on "admin/config/stanford/courses/tag-translate"
+    Then I click on the element with xpath "//table//a[1]"
+    Then I should be on "admin/config/stanford/courses/tag-translate/edit/1"
```
Comment (jbickar) [inline includes/features/stanford_courses_tag_translate.feature]: This fails with:
```
  Then I should be on "admin/config/stanford/courses/tag-translate/edit/1" # SWSMinkContext::assertPageAddress()
      Current page is "/admin/config/stanford/courses/tag-translate/edit/3", but "/admin/config/stanford/courses/tag-translate/edit/1" expected. (Behat\Mink\Exception\ExpectationException)
```

File: jsa/features/stanford_courses_tag_translate.feature [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    Then I should be on "admin/config/stanford/courses/tag-translate"
+    Then I should not see "zzzzzzzZZZZZzzzzzz"
+
+  # This scenario only works if the import scenario about has been run
+  @api @content @dev @deploy
+  Scenario: Edit tag translations
+    Given I am logged in as a user with the "administrator" role
+    And I am on "admin/config/stanford/courses/tag-translate"
+    Then I click on the element with xpath "//table//a[1]"
+    Then I should be on "admin/config/stanford/courses/tag-translate/edit/1"
```
Comment (jbickar) [inline jsa/features/stanford_courses_tag_translate.feature]: Sorry, commented on wrong line.

This fails with:

```
  Then I should be on "admin/config/stanford/courses/tag-translate/edit/1" # SWSMinkContext::assertPageAddress()
      Current page is "/admin/config/stanford/courses/tag-translate/edit/3", but "/admin/config/stanford/courses/tag-translate/edit/1" expected. (Behat\Mink\Exception\ExpectationException)
```

**Comment by jbickar** [comment]
I can't get the Feature in `includes/stanford_courses.feature` to run at all.

When I reference it directly, or symlink it, I get a `No specifications found...` error:

```
spr-jbickar: ~/Documents/working/linky_clicky/jsa$ behat -p local -s all ../includes/features/stanford_courses.feature 
No scenarios
No steps
0m0.07s (14.47Mb)


  [Behat\Testwork\Tester\Exception\WrongPathsException]                                
  No specifications found at path(s) `../includes/features/stanford_courses.feature`.  
```

When I copy it into my `features` directory, I get:

`Fatal error: Call to undefined method Drupal\Driver\BlackboxDriver::drush() in /Users/jbickar/Documents/working/linky_clicky/includes/bootstrap/SWSFeatureContext.php on line 76`

File: jsa/features/stanford_courses_tag_translate.feature (hunk 2 above)
Comment (sherakama) [inline jsa/features/stanford_courses_tag_translate.feature]: I don't seem to be able to count... I thought I had that one nailed down. I am trying to ensure that only the items I want in the test are there. I wonder if their ids change.

**Comment by sherakama** [comment]
`No specifications found at path(s)` is the error that we get when the symlink is not right.
The fatal error looks like your version of behat is different from what is required. Let me adjust the one that is bundled in this repo and run off of it. eg: [repo]/bin/behat

**Comment by jbickar** [comment]
both my install of behat and the one in `bin/` are reporting `behat version 3.1.0rc1`

**Comment by sherakama** [comment]
That is newer than what I have. `behat version 3.0.15`
Might be one of the other dependencies... I will dig in to it.

**Comment by sherakama** [comment]
I think I have figured it out. It was missing the pull request I just merged that added the paths to the config in /includes/config/default.yml

I am going to update composer to include the patched version so that the symlink works. Look for that PR as well.

**Comment by sherakama** [comment]
Take a look at the composer file changes in https://github.com/SU-SWS/linky_clicky/pull/59

I should have separated that out in to its own pull request but hey, past Shea is a jerk.

**Comment by cjwest** [comment]
@sherakama , Would it make sense to have some @deploy tags? Some of these tests would only make sense when the site is first deployed.

**Comment by cjwest** [comment]
Otherwise, it looks good!

8 scenarios (8 passed)
63 steps (63 passed)
3m56.67s (17.42Mb)

**Pull request merged**